docs/architecture/adr-021-abci-events.md

```diff
@@ -0,0 +1,47 @@
+# ADR 012: ABCI Events
+
+## Changelog
+
+- *2018-09-02* Remove ABCI errors component. Update description for events
+- *2018-07-12* Initial version
+
+## Context
+
+ABCI tags were first described in [ADR 002](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-002-event-subscription.md).
+They are key-value pairs that can be used to index transactions.
+
+Currently, ABCI messages return a list of tags to describe an
+"event" that took place during the Check/DeliverTx/Begin/EndBlock,
+where each tag refers to a different property of the event, like the sending and receiving account addresses.
+
+Since there is only one list of tags, recording data for multiple such events in
+a single Check/DeliverTx/Begin/EndBlock must be done using prefixes in the key
```

Comment:
requires extra convention in the keys. not necessarily a bad thing, but could be nicer to have native support for the concept

Reply:
If we can't do with list of `key: value` strings, we should consider bringing in complex types (there was a discussion about this some time ago with Jae where we decided to stick with strings as raw bytes).

If we do so, we can add JSON type (e.g. `key: "Joth", value_type: JSON, value: {"address": "X", "denom": "B"}`). Then it would be super easy to support multiple events per DeliverTx.

Reply:
> there was a discussion about this some time ago with Jae where we decided to stick with strings as raw bytes

Do you know if the reasoning for this was documented anywhere? Didn't we explicitly get rid of types here at one point?

> we can add JSON type

This is cool, but I think it means we'd need to unmarshal the JSON and index the keys in there too if we want to support eg. a query for "validator X slashed". More likely we need to really start thinking about the "indexing outside of tendermint core" story.

Reply:
Let's move this to https://github.com/tendermint/tendermint/issues/1859